File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.py
class Solution:
    def arrayRankTransform(self, arr: List[int]) -> List[int]:
        # heap -> element, idx

        minHeap = []
        for i, num in enumerate(arr):
            heapq.heappush(minHeap, (num, i))

        prev, rank = float(-inf), 0
        while minHeap:
            num, i = heapq.heappop(minHeap)
            if num > prev:
                rank += 1
            arr[i] = rank
            prev = num

        return arr

File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.java
class Solution {
    public int[] arrayRankTransform(int[] arr) {
        TreeMap<Integer, Integer> rankMap = new TreeMap<>();

        for (int num : arr) {
            rankMap.put(num, 0);
        }
        int rank = 1;
        for (int num : rankMap.keySet()) {
            rankMap.put(num, rank++);
        }

        return Arrays.stream(arr).map(rankMap::get).toArray();
    }
}